File: issuer-list/src/main/java/com/adyen/checkout/issuerlist/IssuerListComponent.java
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 25/4/2019.
 */

package com.adyen.checkout.issuerlist;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.SavedStateHandle;

import com.adyen.checkout.components.GenericComponentState;
import com.adyen.checkout.components.base.BasePaymentComponent;
import com.adyen.checkout.components.base.GenericPaymentMethodDelegate;
import com.adyen.checkout.components.model.paymentmethods.InputDetail;
import com.adyen.checkout.components.model.paymentmethods.Issuer;
import com.adyen.checkout.components.model.paymentmethods.Item;
import com.adyen.checkout.components.model.paymentmethods.PaymentMethod;
import com.adyen.checkout.components.model.payments.request.IssuerListPaymentMethod;
import com.adyen.checkout.components.model.payments.request.PaymentComponentData;

import java.util.ArrayList;
import java.util.List;

public abstract class IssuerListComponent<IssuerListPaymentMethodT extends IssuerListPaymentMethod>
        extends BasePaymentComponent<
        IssuerListConfiguration,
        IssuerListInputData,
        IssuerListOutputData,
        GenericComponentState<IssuerListPaymentMethodT>
        > {

    private final MutableLiveData<List<IssuerModel>> mIssuersLiveData = new MutableLiveData<>();

    @SuppressWarnings("LambdaLast")
    public IssuerListComponent(
            @NonNull SavedStateHandle savedStateHandle,
            @NonNull GenericPaymentMethodDelegate genericPaymentMethodDelegate,
            @NonNull IssuerListConfiguration configuration
    ) {
        super(savedStateHandle, genericPaymentMethodDelegate, configuration);
        initComponent(genericPaymentMethodDelegate.getPaymentMethod());
    }

    MutableLiveData<List<IssuerModel>> getIssuersLiveData() {
        return mIssuersLiveData;
    }

    private void initComponent(@NonNull PaymentMethod paymentMethod) {
        final List<Issuer> issuersList = paymentMethod.getIssuers();
        if (issuersList != null) {
            initIssuers(issuersList);
        } else {
            initLegacyIssuers(paymentMethod.getDetails());
        }
    }

    private void initIssuers(@NonNull List<Issuer> issuerList) {
        final List<IssuerModel> issuerModelList = new ArrayList<>();
        for (Issuer issuer : issuerList) {
            if (!issuer.isDisabled()) {
                final IssuerModel issuerModel = new IssuerModel(issuer.getId(), issuer.getName());
                issuerModelList.add(issuerModel);
            }
        }
        mIssuersLiveData.setValue(issuerModelList);
    }

    private void initLegacyIssuers(@Nullable List<InputDetail> details) {
        if (details != null) {
            for (InputDetail detail : details) {
                if (detail.getItems() != null) {
                    final List<IssuerModel> issuers = new ArrayList<>();
                    for (Item item : detail.getItems()) {
                        final IssuerModel issuer = new IssuerModel(item.getId(), item.getName());
                        issuers.add(issuer);
                    }
                    mIssuersLiveData.setValue(issuers);
                }
            }
        }
    }

    @Override
    @NonNull
    protected IssuerListOutputData onInputDataChanged(@NonNull IssuerListInputData inputData) {
        // can also reuse instance if we implement equals properly
        return new IssuerListOutputData(inputData.getSelectedIssuer());
    }

    @NonNull
    @Override
    protected GenericComponentState<IssuerListPaymentMethodT> createComponentState() {
        final IssuerListPaymentMethodT issuerListPaymentMethod = instantiateTypedPaymentMethod();

        final IssuerModel selectedIssuer = getOutputData() != null ? getOutputData().getSelectedIssuer() : null;

        issuerListPaymentMethod.setType(mPaymentMethodDelegate.getPaymentMethodType());
        issuerListPaymentMethod.setIssuer(selectedIssuer != null ? selectedIssuer.getId() : "");

        final boolean isInputValid = getOutputData().isValid();

        final PaymentComponentData<IssuerListPaymentMethodT> paymentComponentData = new PaymentComponentData<>();
        paymentComponentData.setPaymentMethod(issuerListPaymentMethod);

        return new GenericComponentState<>(paymentComponentData, isInputValid, true);
    }

    @NonNull
    protected abstract IssuerListPaymentMethodT instantiateTypedPaymentMethod();

    @NonNull
    protected String getPaymentMethodType() {
        return mPaymentMethodDelegate.getPaymentMethodType();
    }
}

File: issuer-list/src/main/java/com/adyen/checkout/issuerlist/IssuerListComponent.kt
/*
 * Copyright (c) 2019 Adyen N.V.
 *
 * This file is open source and available under the MIT license. See the LICENSE file for more info.
 *
 * Created by caiof on 25/4/2019.
 */
package com.adyen.checkout.issuerlist

import androidx.annotation.RestrictTo
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.adyen.checkout.action.ActionHandlingComponent
import com.adyen.checkout.action.DefaultActionHandlingComponent
import com.adyen.checkout.action.GenericActionDelegate
import com.adyen.checkout.components.ButtonComponent
import com.adyen.checkout.components.PaymentComponent
import com.adyen.checkout.components.PaymentComponentEvent
import com.adyen.checkout.components.PaymentComponentState
import com.adyen.checkout.components.base.ComponentDelegate
import com.adyen.checkout.components.base.ComponentEventHandler
import com.adyen.checkout.components.extensions.mergeViewFlows
import com.adyen.checkout.components.model.payments.request.IssuerListPaymentMethod
import com.adyen.checkout.components.toActionCallback
import com.adyen.checkout.components.ui.ButtonDelegate
import com.adyen.checkout.components.ui.ViewableComponent
import com.adyen.checkout.components.ui.view.ComponentViewType
import com.adyen.checkout.core.log.LogUtil
import com.adyen.checkout.core.log.Logger
import kotlinx.coroutines.flow.Flow

/**
 * Component should not be instantiated directly.
 */
abstract class IssuerListComponent<IssuerListPaymentMethodT : IssuerListPaymentMethod> protected constructor(
    private val issuerListDelegate: IssuerListDelegate<IssuerListPaymentMethodT>,
    private val genericActionDelegate: GenericActionDelegate,
    private val actionHandlingComponent: DefaultActionHandlingComponent,
    @get:RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    val componentEventHandler: ComponentEventHandler<PaymentComponentState<IssuerListPaymentMethodT>>,
) : ViewModel(),
    PaymentComponent,
    ViewableComponent,
    ButtonComponent,
    ActionHandlingComponent by actionHandlingComponent {

    final override val delegate: ComponentDelegate get() = actionHandlingComponent.activeDelegate

    override val viewFlow: Flow<ComponentViewType?> = mergeViewFlows(
        viewModelScope,
        issuerListDelegate.viewFlow,
        genericActionDelegate.viewFlow,
    )

    init {
        issuerListDelegate.initialize(viewModelScope)
        genericActionDelegate.initialize(viewModelScope)
        componentEventHandler.initialize(viewModelScope)
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    fun observe(
        lifecycleOwner: LifecycleOwner,
        callback: (PaymentComponentEvent<PaymentComponentState<IssuerListPaymentMethodT>>) -> Unit
    ) {
        issuerListDelegate.observe(lifecycleOwner, viewModelScope, callback)
        genericActionDelegate.observe(lifecycleOwner, viewModelScope, callback.toActionCallback())
    }

    @RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
    fun removeObserver() {
        issuerListDelegate.removeObserver()
        genericActionDelegate.removeObserver()
    }

    override fun isConfirmationRequired(): Boolean = issuerListDelegate.isConfirmationRequired()

    override fun submit() {
        (delegate as? ButtonDelegate)?.onSubmit() ?: Logger.e(TAG, "Component is currently not submittable, ignoring.")
    }

    override fun setInteractionBlocked(isInteractionBlocked: Boolean) {
        (delegate as? IssuerListDelegate<*>)?.setInteractionBlocked(isInteractionBlocked)
            ?: Logger.e(TAG, "Payment component is not interactable, ignoring.")
    }

    override fun onCleared() {
        super.onCleared()
        Logger.d(TAG, "onCleared")
        issuerListDelegate.onCleared()
        genericActionDelegate.onCleared()
        componentEventHandler.onCleared()
    }

    companion object {
        private val TAG = LogUtil.getTag()
    }
}
